incubator/hnc/test/conversion/conversion_test.go

```diff
@@ -0,0 +1,144 @@
+
+	. "sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils"
+)
+
+const namspacePrefix = "e2e-conversion-test-"
+
+func TestConversion(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	SetDefaultEventuallyTimeout(time.Second * 2)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"HNC v1alpha2 Conversion Test Suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = Describe("Conversion from v1alpha1 to v1alpha2", func() {
+	const (
+		// hncVersion has to be a version using v1alpha1.
+		hncVersion = "v0.5.0"
+		nsA    = namspacePrefix + "a"
+		nsB    = namspacePrefix + "b"
+		nsC    = namspacePrefix + "c"
+	)
+
+	var (
+		crds = []string{"subnamespaceanchors.hnc.x-k8s.io","hierarchyconfigurations.hnc.x-k8s.io", "hncconfigurations.hnc.x-k8s.io"}
+	)
+
+	BeforeEach(func() {
+		// We need GOPATH here to install kubectl-hns plugin.
+		CheckGoPath()
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Tear down HNC for both the specified version and at HEAD.
+		TearDownHNC(hncVersion)
+		// Wait for HNC to be fully torn down.
+		ensureHNCCleared(crds)
+	})
+
+	AfterEach(func() {
+		CleanupNamespaces(nsA, nsB, nsC)
+		// Only tear down HNC at HEAD, since that's what we just deployed.
+		TearDownHNC("")
+	})
+
+	It("should convert CRD versions properly and HNC should still work", func() {
+		// Set up v1alpha1 and verify everything is fine before conversion.
+		GinkgoT().Log("Set up v1alpha1")
+		setupV1alpha1(hncVersion)
+		// Wait for the validating webhooks to be ready.
+		ensureVWHReady()
+		// Create namespace 'a' and a subnamespace 'b'
+		GinkgoT().Log("Apply v1alpha1 resources")
+		MustRun("kubectl apply -f v1alpha1_resources.yaml")
+		// Create a namespace 'c' and use kubectl plugin to set its parent to 'a'
+		MustRun("kubectl create ns", nsC)
+		GinkgoT().Log("Use v1alpha1 kubectl-hns plugin")
+		MustRun("kubectl hns set", nsC, "--parent", nsA)
+
+		// Verify kubectl-hns plugin works and 'b' and 'c' are children of 'a'.
+		RunShouldContainMultiple([]string{nsB, nsC}, 2, "kubectl hns tree", nsA)
+		// Verify there's no 'v1alpha2' in all three CRDs for now.
+		checkCRDVersionInField("v1alpha2", ".spec", crds, false,2)
+		checkCRDVersionInField("v1alpha2", ".status", crds, false,2)
+
+		// Deploy the HEAD to include API conversion and install v1alpha2 kubectl plugin.
+		GinkgoT().Log("Set up v1alpha2")
+		setupV1alpha2()
+		// Wait for the cert rotator to write caBundles in CRD conversion webhooks.
+		ensureCRDConvWHReady(crds)
+
+		// Verify the conversion works with 'v1alpha1' still in the spec.versions but not in status.storedVersions.
+		checkCRDVersionInField("v1alpha1", ".spec.versions", crds, true, 2)
+		checkCRDVersionInField("v1alpha2", ".spec.versions", crds, true, 2)
+		// Give this a longer timeout for the reconciler to remove v1alpha1 from storedVersions.
+		checkCRDVersionInField("v1alpha1", ".status.storedVersions", crds, false, 5)
+		checkCRDVersionInField("v1alpha2", ".status.storedVersions", crds, true,2)
```

Comment:
Maybe always use a 5s timeout and remove it from the list of parameters? It's a little confusing to read and waiting an additional 3s for a (rare) failure seems fine to me.